File: fenwick-tree/csharp/1850.cs
using System;
namespace _1850___Again_Making_Queries_II
{
    class Program
    {
        static int MOD = 10000;
        static void Main(string[] args)
        {
            string[] data = Console.ReadLine().Split();
            int N = int.Parse(data[0]);
            int U = int.Parse(data[1]);
            int Q = int.Parse(data[2]);

            Fenwick_Tree ft = new Fenwick_Tree(N);

            for (int i = 0; i < U + Q; i++)
            {
                data = Console.ReadLine().Split();
                int op = int.Parse(data[0]);
                int u = int.Parse(data[1]);
                int v = int.Parse(data[2]);

                if (op == 1)
                {
                    ft.update(u, 2 * v);
                    if (u != 1)
                        ft.update(u - 1, v);
                    if (u != N)
                        ft.update(u + 1, v);
                }
                else
                {
                    Console.WriteLine(ft.query(u, v));
                }
            }
        }
        private class Fenwick_Tree
        {
            private int _n;
            private int[] ft;

            public Fenwick_Tree(int n)
            {
                _n = n;
                ft = new int[_n + 1];
            }
            private int lowbit(int v)
            {
                return v & (-v);
            }
            internal long query(int u, int v)
            {
                long sum = query(v) - query(u - 1);
                if (sum < 0)
                    sum += MOD;
                return sum;
            }
            private long query(int pos)
            {
                long sum = 0;
                for (; pos > 0; pos -= lowbit(pos))
                    sum = (sum + ft[pos]) % MOD;
                return sum;
            }
            internal void update(int pos, int val)
            {
                for (; pos <= _n; pos += lowbit(pos))
                    ft[pos] = (ft[pos] + val) % MOD;
            }
        }
    }
}

File: fenwick-tree/cplusplus/1850.cpp
#include <bits/stdc++.h>

#define endl '\n'
#define MOD 10000

using namespace std;

typedef long long Int;

template <class T>
struct BIT
{

  vector<T> ft;
  BIT(int n) : ft(n, 0) {}

  void Update(int pos, int val)
  {
    for (; pos <= ft.size(); pos += pos & (-pos))
      ft[pos] = (ft[pos] + val) % MOD;
  }

  Int Query(int pos)
  {
    Int sum = 0;

    for (; pos > 0; pos -= pos & (-pos))
      sum = (sum + ft[pos]) % MOD;
    return sum;
  }

  Int Query(int a, int b)
  {
    Int sum = Query(b) - Query(a - 1);

    if (sum < 0)
      sum += MOD;
    return sum;
  }
};

int main()
{
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int N, U, Q;
  cin >> N >> U >> Q;
  BIT<int> A(N + 1);

  for (int i = 0; i < U + Q; i++)
  {
    int op, u, v;
    cin >> op >> u >> v;

    if (op == 1)
    {
      A.Update(u, 2 * v);
      if (u != 1)
        A.Update(u - 1, v);
      if (u != N)
        A.Update(u + 1, v);
    }

    else
    {
      cout << A.Query(u, v) << endl;
    }
  }

  return 0;
}
